### PR TITLE
StyleGAN Generative Model for ADNI Brain Dataset with t-SNE Visualization

### PR DESCRIPTION
**Pull request by PuttiKoselig**
This project implements a generative adversarial network (GAN) architecture, specifically a modified version of StyleGAN, to synthesize high-quality brain images from the Alzheimer’s Disease Neuroimaging Initiative (ADNI) dataset (https://adni.loni.usc.edu/). The model employs various layers such as Pixel Normalization, Adaptive Instance Normalization, and Weighted Sum Linear layers to effectively generate images that mimic real brain scans.

For more detailed information about the model, its performance, and any dependencies, please see the README file. The github commit log shows the progress and development of the algorithm. All files related to the model, including the README, are in the directory location recognition/StyleGAN_VBV.

The files added are:

- **modules.py:** Model components: Generator and Discriminator
- **dataset.py:** Data loading and preprocessing
- **train.py:** Training, validation, and saving the model
- **predict.py:** Usage example for generating images
- **embedding.py:** Script for t-SNE embedding visualization
- **README.md:** Project documentation
- images and plots demonstrating the code functionality can be found in the 'images' subdirectory

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (4.5 Marks)
Clear generated images;
Embedding somehow not clustering -0.5

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Modularized design

5. Commenting: (1 Mark)
Sufficient docstrings and comments

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- The intention of this task is actually training one single model on the AD and NC data combined, and plotting the embedding on the style space so that clustering of two classes is expected in the latent space (see the reference paper). The way you did it is essentially similar but more computationally expensive, which is fine. However, there's no clustering/separation in your embedding, likely because you are plotting the combined embedding by `plot_tsne()` rather than firstly splitting the embedding based on AD/NC and then plotting them together;
- It's a good idea to plot the G loss and D loss of one model together, since it's adversarial training, we'd like to see how the two losses fight against each other;
- The `predict.py` seems not updated, the functions declared not been used/called;
- The description under Trainning Process on README.md could be enriched.

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: 
- Loss figures are too small. 
- The loss of all these figures increased after certain epochs; it might be gradient problems, and early stopping might be required. 
- The train.py indicates it is a progressive growing of image size. If so, the training loss would be too smooth. Or is the shown loss only from one training stage?
- T-SNE embedding was generated from two models but still could not see clear clusters.
-  It would be better to plot generated images at each resolution stage.
2. **Implementation**: Data augmentation was applied. Hyperparameters were hard-coded. Otherwise, well structured.
3. **Design**: Good
4. **Comment**: Sufficient
5. **Difficulty**: Hard

**Note**
pycache and DS_store should not be committed.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Model has some issues.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. Abstract comments.</td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No references.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Remove cache files for merge.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-6</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by PuttiKoselig** [comment]
Hi, 
I am not entirely clear which changes are the ones specifically requested to obtain the +2 feedback marks. Is it just to remove the cache files for merge listed besides "feedback action required"?

Thank you

**Comment by PuttiKoselig** [comment]
Thank you so much for the feedback, I have removed the cache files as requested.

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.

**Comment by PuttiKoselig** [comment]
Hi, What do I need to change in particular? I'm not entirely clear on whats causing the merge problem.

**Comment by PuttiKoselig** [comment]
Hi Shakes, I think I've fixed this now. Please let me know if there is anything else I need to do.

**Comment by shakes76** [comment]
Feedback applied +2
